controllers/etcd_custodian_controller.go

```diff
@@ -0,0 +1,145 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	druidv1alpha1 "github.com/gardener/etcd-druid/api/v1alpha1"
+	"github.com/gardener/gardener/pkg/utils/kubernetes/health"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+// EtcdCustodian reconciles status of Etcd object
+type EtcdCustodian struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// NewEtcdCustodian creates a new EtcdCustodian object
+func NewEtcdCustodian(mgr manager.Manager) *EtcdCustodian {
+	return &EtcdCustodian{
+		Client: mgr.GetClient(),
+		Scheme: mgr.GetScheme(),
+	}
+}
+
+// +kubebuilder:rbac:groups=druid.gardener.cloud,resources=etcds,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=druid.gardener.cloud,resources=etcds/status,verbs=get;update;patch
+
+// Reconcile reconciles the etcd.
+func (ec *EtcdCustodian) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.TODO()
+	etcd := &druidv1alpha1.Etcd{}
+	if err := ec.Get(ctx, req.NamespacedName, etcd); err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return. Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// If any adoptions are attempted, we should first recheck for deletion with
+	// an uncached quorum read sometime after listing Machines (see #42639).
+	canAdoptFunc := RecheckDeletionTimestamp(func() (metav1.Object, error) {
+		foundEtcd := &druidv1alpha1.Etcd{}
+		err := ec.Get(context.TODO(), types.NamespacedName{Name: etcd.Name, Namespace: etcd.Namespace}, foundEtcd)
+		if err != nil {
+			return nil, err
+		}
+		if foundEtcd.UID != etcd.UID {
+			return nil, fmt.Errorf("original %v/%v hvpa gone: got uid %v, wanted %v", etcd.Namespace, etcd.Name, foundEtcd.UID, etcd.UID)
+		}
+		return foundEtcd, nil
+	})
+
+	selector, err := metav1.LabelSelectorAsSelector(etcd.Spec.Selector)
+	if err != nil {
+		logger.Error(err, "Error converting etcd selector to selector")
+		return ctrl.Result{}, err
+	}
+
+	cm := NewEtcdDruidRefManager(ec.Client, ec.Scheme, etcd, selector, etcdGVK, canAdoptFunc)
+
+	ss, err := cm.FetchStatefulSet(etcd)
+	if err != nil {
+		return ctrl.Result{
+			Requeue: true,
+		}, err
+	}
+
+	// If no statefulsets could be fetched, requeue for reconcilation
+	if len(ss) < 1 {
+		return ctrl.Result{
+			Requeue: true,
```

Comment:
Requeue after a fixed delay might be better because it requires the main reconciliation to happen and there is no point in requeuing aggressively here.

Reply:
Excellent observation. Thanks.

Reply:
Thanks for change. But I meant to use the [`RequeueAfter` field in `Result`](https://github.com/gardener/etcd-druid/blob/8307a62ad71ec0de3bf78de6d0e2bbc07700042c/vendor/sigs.k8s.io/controller-runtime/pkg/reconcile/reconcile.go#L32) rather than a synchronous sleep.

Reply:
Done @amshuman-kr